internal/resource-request-operator/resourceRequest_controller.go

```diff
@@ -0,0 +1,162 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/util/slice"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"time"
+)
+
+// ResourceRequestReconciler reconciles a ResourceRequest object
+type ResourceRequestReconciler struct {
+	client.Client
+	Log       logr.Logger
+	Scheme    *runtime.Scheme
+	ClusterId string
+}
+
+type ResourceToOffer struct {
+	Offers corev1.ResourceList
+}
+
+var resources ResourceToOffer
+
+const (
+	finalizerName = "discovery.liqo.io.resourceRequest.finalizer"
+	offerPrefix   = "resourceoffer-"
+)
+
+// +kubebuilder:rbac:groups=discovery.liqo.io,resources=resourceRequests,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=discovery.liqo.io,resources=resourceRequests/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=sharing.liqo.io,resources=resourceOffers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=sharing.liqo.io,resources=resourceOffers/status,verbs=get;update;patch
+
+func (r *ResourceRequestReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	var resourceRequest discoveryv1alpha1.ResourceRequest
+	err := r.Get(ctx, req.NamespacedName, &resourceRequest)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		} else {
+			klog.Errorf("%s -> unable to get resourceRequest %s: %s", r.ClusterId, req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+	}
+	//not in delete
+	if resourceRequest.DeletionTimestamp.IsZero() {
+		// add finalizer
+		if !slice.ContainsString(resourceRequest.GetFinalizers(), finalizerName, nil) {
+			resourceRequest.SetFinalizers(append(resourceRequest.GetFinalizers(), finalizerName))
+			if err := r.Update(ctx, &resourceRequest); err != nil {
+				klog.Errorf("%s -> unable to update resource finalizer %s: %s", r.ClusterId, req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+
+		}
+	} else {
+		//the resource is being deleted
+		if slice.ContainsString(resourceRequest.GetFinalizers(), finalizerName, nil) {
+
+			// remove our finalizer from the list and update it.
+			resourceRequest.SetFinalizers(slice.RemoveString(resourceRequest.GetFinalizers(), finalizerName, nil))
+			if err := r.Update(ctx, &resourceRequest); err != nil {
+				klog.Errorf("%s -> unable to update resource finalizer %s: %s", r.ClusterId, req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	offer, offerErr := r.generateResourceOffer(&resourceRequest)
+	if offerErr != nil {
+		if offer != nil {
+			klog.Errorf("%s -> unable to generate resourceOffer %s", r.ClusterId, offer.Name)
+		}
+		klog.Errorf("%s -> Error generating resourceOffer: %s", r.ClusterId, offerErr)
```

Comment:
AFAIK the typical behavior is to return nil in case of error, and != nil if no error occurred